index.js

```diff
@@ -1,205 +1,47 @@
-    }
-
-    e.statusCode = e.statusCode || 500
-    throw e
-  }
-}
-
-function getSecret(request, reply, cb) {
-  request
-    .jwtDecode({ decode: { complete: true } })
-    .then(decoded => {
-      const { header } = decoded
-
-      // If the algorithm is not using RS256, the encryption key is Auth0 client secret
-      if (header.alg.startsWith('HS')) {
-        return cb(null, request.auth0Verify.secret)
-      }
-
-      // If the algorithm is RS256, get the key remotely using a well-known URL containing a JWK set
-      getRemoteSecret(request.auth0Verify.domain, header.alg, header.kid, request.auth0VerifySecretsCache)
-        .then(key => cb(null, key))
-        .catch(cb)
-    })
-    .catch(cb)
-}
-
-async function authenticate(request, reply) {
-  try {
-    await request.jwtVerify()
-  } catch (e) {
-    for (const [jwtMessage, errorMessage] of fastifyJwtErrors) {
-      if (e.message.match(jwtMessage)) {
-        throw new Unauthorized(errorMessage, { a: 1 })
+      if (!domain.match(/^http(?:s?)/)) {
+        domain = new URL(`https://${domain}`).toString()
+      } else {
+        // Add missing trailing slash if it hasn't been provided in the config
+        domain = new URL(domain).toString()
       }
-    }
 
-    if (e.statusCode) {
-      throw e
+      options.jwksUrl = `${domain}.well-known/jwks.json`
+    } else if (!secret) {
+      // Throw missing options error here to prevent confusing error
+      // message from fastify-jwt-jwks being thrown to user
+      throw new Error(errorMessages.missingOptions)
     }
 
-    throw new Unauthorized(e.message)
-  }
-}
-
-function fastifyAuth0Verify(instance, options, done) {
-  try {
-    // Check if secrets cache is wanted - Convert milliseconds to seconds and cache for a week by default
-    const ttl = parseFloat('secretsTtl' in options ? options.secretsTtl : '604800000', 10) / 1e3
-    delete options.secretsTtl
+    const plugin = fastifyJwtJwks(instance, options, done)
 
-    const auth0Options = verifyOptions(options)
-
-    // Setup @fastify/jwt
-    instance.register(fastifyJwt, {
-      verify: auth0Options.verify,
-      cookie: options.cookie,
-      secret: getSecret,
-      jwtDecode: 'jwtDecode',
-      formatUser: options.formatUser
-    })
-
-    // Setup our decorators
-    instance.decorate('authenticate', authenticate)
-    instance.decorate('auth0Verify', auth0Options)
-    instance.decorateRequest('auth0Verify', {
-      getter: () => auth0Options
-    })
-
-    const cache =
-      ttl > 0 ? new NodeCache({ stdTTL: ttl }) : { get: () => undefined, set: () => false, close: () => undefined }
+    if (!instance.jwtJwks) {
+      throw new Error('Big scary error')
+    }
 
-    // Create a cache or a fake cache
-    instance.decorateRequest('auth0VerifySecretsCache', {
-      getter: () => cache
+    instance.addHook('onReady', function (done) {
+      instance.decorate('auth0Verify', instance.jwtJwks)
+      done()
     })
```

Comment:
I don't like much this implementation either. We can register our decorator when the callback of the other plugins is called, meaning that the other plugin has initialized and the decorator should be available.

Please watch out for binding issues in providing instance.jwtJwks directly in the decorator initialization.

I understand tests are passing but I wouldn't be too sure they're testing explicitly for these scenarios, as these are potentially novel issues related to the implementation change